File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-8.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    private int preOrder(TreeNode node, int ancestorMax){
        if (node == null) return 0;

        int res = (node.val >= ancestorMax) ? 1 : 0;
        ancestorMax = Math.max(ancestorMax, node.val);
        return res + preOrder(node.left, ancestorMax) + preOrder(node.right, ancestorMax);
    }
    public int goodNodes(TreeNode root) {
        return preOrder(root, root.val);
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-8.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    private bool DFS(TreeNode root, int minVal, int maxVal){
        if (root == null) return true;

        if (root.val <= minVal || root.val >= maxVal){
            return false;
        }

        return DFS(root.left, minVal, root.val) && DFS(root.right, root.val, maxVal);
    }
    public bool IsValidBST(TreeNode root) {
        return DFS(root, int.MinValue, int.MaxValue);
    }
}
